app/src/main/java/de/westnordost/streetcomplete/quests/surface/AddCyclewayPartSurface.kt

```diff
@@ -25,7 +25,11 @@ class AddCyclewayPartSurface : OsmFilterQuestType<SurfaceAnswer>() {
             and !note:cycleway:surface
           )
         )
-        and (access !~ private|no or (foot and foot !~ private|no) or (bicycle and bicycle !~ private|no))
+        and (
+          !access and bicycle !~ private|no
+          or access and access !~ private|no and bicycle !~ private|no
+          or access and access  ~ private|no and bicycle and bicycle !~ private|no
+        )
```

Comment:
Why this?

Reply:
This addresses second bullet point of this comment: https://github.com/streetcomplete/StreetComplete/pull/4548#issuecomment-1290703523

Basically, the access logic didn't work correctly in original code (i.e. pre-this PR).
For example, the `cycleway surface` was still being asked for `bicycle=private` etc. (which, as I understand it, shouldn't happen).

So I've written in way that is:
- actually working correctly (passes the tests that check for all combinations)
- more readable (IMHO), as it checks each general `access=*` state separately on each line
- (taking a little more space)

Reply:
> quests in general are asked for ways that are either accessible by default, on foot or with the vehicle type you ought to use there

I agree (and that is what I've tried to do), but the English sentence above have the similary problem the code had; and it is interaction between those **cannot** be defined with simple **"or"** (e.g. `foot=no` overrides `access=yes`).

> How much of a theoretical case is this, though?

That I don't know (and I worry it is unlikely that ad-hoc overpass queries trying to find out would survive such attempts on word-wide queries)


> I am still struggling what exactly the cited section of code filters.... would it not evaluate to false for a way that has access=no but which you can access on foot?

It would evaluate to `false`, and correctly IMHO, because `foot` does not matter at all in `cycleway surface` quest. (You qouted `AddCyclewayPartSurface.kt` here!). There is very similar (but with key names changed) code in `AddFootwayPartSurface.kt` where your example of `access=no+foot=yes` would evaluate to `true`:

```
          !access and foot !~ private|no
          or access and access !~ private|no and foot !~ private|no
          or access and access  ~ private|no and foot and foot !~ private|no
```

The new logic seems quite readable to me:
- if general `access` is not defined, check if `foot` specifically is disallowed
- if general `access` is allowed, check if `foot` overrides this
- if general `access` is disallowed, check if `foot` overrides this

There might be **_shorter_** ways to write this, but much **_less readable_** (in my opinion). I found original code not only give incorrect results, but was also treacherous/misleading and confusing (partly due to how SC handles `null` values. YMMV)

> Also, for consistency, such change necessitates also changing dozens of other quest filters to be consistent (hence my initial knee-jerk reaction that this should be an own PR, as it has the potential to grow out of control).

Well I don't disagree that it should be solved globally, but would rather not assign such a task to _myself_ at this time. I've added the automated tests because you've asked me to, and then tried to handle situation when they failed (by fixing incorrect quest code). 

---

Anyway, let me know how you would prefer this to be resolved. IMHO, the current solution is the best one (I'm obviously biased), but if you want me to revert to old access-handling code and remove/comment-out the tests finding inconsistencies (or some other solution), let me know and I'll try to oblige.

Reply:
I prefer to not have that in this PR then. If this is done, it should be done consistently across all quests where such a filter or a similar filter exists. But I am not at all convinced about either that it is supposedly easier to read, that it covers a relevant case or even that the logic would be correct.  As written in my last post.

Reply:
OK, done as requested (reverted to old access checking code, and commented out the tests that would be failing)